File: ios/LivenessDetector.swift
import Foundation

class LivenessDetector {
  private let BLINK_THRESHOLD: Float = 0.25
  private let SMILE_THRESHOLD: Float = 0.45
  private let HEAD_TURN_THRESHOLD: Float = 15
  private let NOD_THRESHOLD: Float = 12

  func computeEAR(_ eyeLandmarks: [[String: Any]]) -> Float {
    guard eyeLandmarks.count >= 6 else { return 1.0 }

    let points = eyeLandmarks.compactMap { p -> (Float, Float)? in
      guard let x = p["x"] as? Double, let y = p["y"] as? Double else { return nil }
      return (Float(x), Float(y))
    }

    guard points.count >= 6 else { return 1.0 }

    let dist = { (a: (Float, Float), b: (Float, Float)) -> Float in
      sqrt(pow(a.0 - b.0, 2) + pow(a.1 - b.1, 2))
    }

    let vertical1 = dist(points[1], points[5])
    let vertical2 = dist(points[2], points[4])
    let horizontal = dist(points[0], points[3])

    return (vertical1 + vertical2) / (2.0 * horizontal)
  }

  func computeMAR(_ mouthLandmarks: [[String: Any]]) -> Float {
    guard mouthLandmarks.count >= 8 else { return 0.0 }

    let points = mouthLandmarks.compactMap { p -> (Float, Float)? in
      guard let x = p["x"] as? Double, let y = p["y"] as? Double else { return nil }
      return (Float(x), Float(y))
    }

    guard points.count >= 8 else { return 0.0 }

    let dist = { (a: (Float, Float), b: (Float, Float)) -> Float in
      sqrt(pow(a.0 - b.0, 2) + pow(a.1 - b.1, 2))
    }

    let vertical = dist(points[2], points[6])
    let horizontal = dist(points[0], points[4])

    return vertical / horizontal
  }

  func isBlink(_ landmarks: FaceLandmarks) -> Bool {
    let leftEAR = computeEAR(landmarks.leftEye)
    let rightEAR = computeEAR(landmarks.rightEye)
    let avgEAR = (leftEAR + rightEAR) / 2
    return avgEAR < BLINK_THRESHOLD
  }

  func isSmile(_ landmarks: FaceLandmarks) -> Bool {
    let mar = computeMAR(landmarks.mouth)
    return mar > SMILE_THRESHOLD
  }

  func isHeadTurnLeft(_ landmarks: FaceLandmarks) -> Bool {
    let yaw = landmarks.headPose["yaw"] as? Double ?? 0
    return yaw < -Double(HEAD_TURN_THRESHOLD)
  }

  func isHeadTurnRight(_ landmarks: FaceLandmarks) -> Bool {
    let yaw = landmarks.headPose["yaw"] as? Double ?? 0
    return yaw > Double(HEAD_TURN_THRESHOLD)
  }

  func isNod(_ landmarks: FaceLandmarks) -> Bool {
    let pitch = landmarks.headPose["pitch"] as? Double ?? 0
    return abs(pitch) > Double(NOD_THRESHOLD)
  }
}

File: android/app/src/main/java/com/faceauthoffline/LivenessDetector.kt
package com.faceauthoffline

import android.util.Log
import kotlin.math.abs
import kotlin.math.sqrt

class LivenessDetector {
  fun computeEAR(eyeLandmarks: List<Pair<Float, Float>>): Float {
    if (eyeLandmarks.size < 6) return 1.0f
    val dist = { a: Pair<Float, Float>, b: Pair<Float, Float> ->
      sqrt((a.first - b.first) * (a.first - b.first) + (a.second - b.second) * (a.second - b.second))
    }
    val vertical1 = dist(eyeLandmarks[1], eyeLandmarks[5])
    val vertical2 = dist(eyeLandmarks[2], eyeLandmarks[4])
    val horizontal = dist(eyeLandmarks[0], eyeLandmarks[3])
    return (vertical1 + vertical2) / (2.0f * horizontal)
  }

  fun computeMAR(mouthLandmarks: List<Pair<Float, Float>>): Float {
    if (mouthLandmarks.size < 8) return 0.0f
    val dist = { a: Pair<Float, Float>, b: Pair<Float, Float> ->
      sqrt((a.first - b.first) * (a.first - b.first) + (a.second - b.second) * (a.second - b.second))
    }
    val vertical = dist(mouthLandmarks[2], mouthLandmarks[6])
    val horizontal = dist(mouthLandmarks[0], mouthLandmarks[4])
    return vertical / horizontal
  }

  fun isBlink(landmarks: FaceLandmarks): Boolean {
    val leftEAR = computeEAR(landmarks.leftEye)
    val rightEAR = computeEAR(landmarks.rightEye)
    val avgEAR = (leftEAR + rightEAR) / 2
    return avgEAR < BLINK_THRESHOLD
  }

  fun isSmile(landmarks: FaceLandmarks): Boolean {
    val mar = computeMAR(landmarks.mouth)
    return mar > SMILE_THRESHOLD
  }

  fun isHeadTurnLeft(landmarks: FaceLandmarks): Boolean {
    return landmarks.headPose.yaw < -HEAD_TURN_THRESHOLD
  }

  fun isHeadTurnRight(landmarks: FaceLandmarks): Boolean {
    return landmarks.headPose.yaw > HEAD_TURN_THRESHOLD
  }

  fun isNod(landmarks: FaceLandmarks): Boolean {
    return abs(landmarks.headPose.pitch) > NOD_THRESHOLD
  }

  companion object {
    private const val BLINK_THRESHOLD = 0.25f
    private const val SMILE_THRESHOLD = 0.45f
    private const val HEAD_TURN_THRESHOLD = 15f
    private const val NOD_THRESHOLD = 12f
  }
}
